File: MobileApp/Assets/Scripts/Helm.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using UnityEngine;
using UnityEngine.EventSystems;
using UnityEngine.UI;
using static Network;

public class Helm : MonoBehaviour
{

    [Serializable]
    public class ToHelmStation
    {
        public int throttle; // 0-100%
        public int yaw; // -100% - +100%
        public int pitch; // -100% - +100%
        public int enginePower;
        public int jumpPower;
        public int thrusterPower;
    }

    [Serializable]
    public class FromHelmStation
    {
        public int enginePower;
        public int enginePendingPower;
        public int jumpPower;
        public int jumpPendingPower;
        public int thrusterPower;
        public int thrusterPendingPower;
        public int crew;
        public bool crewIsEvac;
        public bool jumpIsAvailable;
        public bool action1IsAvailable;
        public bool action2IsAvailable;
        public bool action3IsAvailable;
        public int maxShields;
        public int foreShields;
        public int aftShields;
        public int portShields;
        public int starShields;
        public int reserve;
        public int maxReserve;
    }

    private Network Network { get; set; }
    private FlightPad FlightPad { get; set; }
    private Slider Throttle { get; set; }
    private EnergyBar EnginePower { get; set; }
    private EnergyBar JumpPower { get; set; }
    private EnergyBar ThrusterPower { get; set; }
    private FourDigit Crew { get; set; }
    private LitButton Jump { get; set; }
    private LitButton Evac { get; set; }
    private ShieldDisplay Shields { get; set; }
    private PowerWheel ReservePower { get; set; }
    private Button Action1 { get; set; }
    private Button Action2 { get; set; }
    private Button Action3 { get; set; }

    private double Elapsed { get; set; }

    private FourDigit PosX { get; set; }
    private FourDigit PosY { get; set; }
    private FourDigit PosZ { get; set; }
    private FourDigit RotX { get; set; }
    private FourDigit RotY { get; set; }

    void Start()
    {

        // get references
        Network = GameObject.Find("Interface").GetComponent<Network>();
        var helm = GameObject.Find("Helm");
        for (int i = 0; i < helm.transform.childCount; i++)
        {
            var obj = helm.transform.GetChild(i).gameObject;
            switch (obj.name)
            {
                case "FlightPad":
                    FlightPad = obj.GetComponent<FlightPad>();
                    break;
                case "Throttle":
                    Throttle = obj.GetComponent<Slider>();
                    break;
                case "EnginePower":
                    EnginePower = obj.GetComponent<EnergyBar>();
                    break;
                case "JumpPower":
                    JumpPower = obj.GetComponent<EnergyBar>();
                    break;
                case "ThrusterPower":
                    ThrusterPower = obj.GetComponent<EnergyBar>();
                    break;
                case "Crew":
                    Crew = obj.GetComponent<FourDigit>();
                    break;
                case "Jump":
                    Jump = obj.GetComponent<LitButton>();
                    break;
                case "Evac":
                    Evac = obj.GetComponent<LitButton>();
                    break;
                case "Shields":
                    Shields = obj.GetComponent<ShieldDisplay>();
                    break;
                case "ReservePower":
                    ReservePower = obj.GetComponent<PowerWheel>();
                    break;
                case "Action1":
                    Action1 = obj.GetComponent<Button>();
                    break;
                case "Action2":
                    Action2 = obj.GetComponent<Button>();
                    break;
                case "Action3":
                    Action3 = obj.GetComponent<Button>();
                    break;
                case "POS-X":
                    PosX = obj.GetComponent<FourDigit>();
                    break;
                case "POS-Y":
                    PosY = obj.GetComponent<FourDigit>();
                    break;
                case "POS-Z":
                    PosZ = obj.GetComponent<FourDigit>();
                    break;
                case "ROT-X":
                    RotX = obj.GetComponent<FourDigit>();
                    break;
                case "ROT-Y":
                    RotY = obj.GetComponent<FourDigit>();
                    break;
            }
        }

    }

    void Update()
    {

        // read interface changes every 200 ms
        Elapsed += Time.deltaTime;
        if (Elapsed > 0.2)
        {
            Elapsed = 0.0f;
            SendToHelmStation();
        }

    }

    public void ReceiveTelemetry(TelemetryPayload payload)
    {

        // set position
        PosX.SetToNumber(Mathf.RoundToInt(payload.posx / 1000.0f));
        PosY.SetToNumber(Mathf.RoundToInt(payload.posy / 1000.0f));
        PosZ.SetToNumber(Mathf.RoundToInt(payload.posz / 1000.0f));

        // x is the y-dim and y is heading
        RotX.SetToNumber(Mathf.RoundToInt(payload.roty));
        var heading = (payload.rotx < 100) ? Mathf.CeilToInt(payload.rotx * -1) : Mathf.CeilToInt(360 - payload.rotx);
        RotY.SetToNumber(heading);

    }

    private void SendToHelmStation()
    {
        ToHelmStation payload = new ToHelmStation();

        // get yaw and pitch
        payload.yaw = FlightPad.Yaw;
        payload.pitch = FlightPad.Pitch;

        // calculate throttle
        payload.throttle = Mathf.CeilToInt(Throttle.value * 100.0f);

        // send desired power
        payload.enginePower = EnginePower.DesiredPowerLevel;
        EnginePower.DesiredPowerLevel = -1; // reset
        payload.jumpPower = JumpPower.DesiredPowerLevel;
        JumpPower.DesiredPowerLevel = -1; // reset
        payload.thrusterPower = ThrusterPower.DesiredPowerLevel;
        ThrusterPower.DesiredPowerLevel = -1; // reset

        // send info
        var msg = new Message<ToHelmStation>()
        {
            c = "helm",
            p = payload,
            e = 0
        };
        Network.Send(msg);

    }

    public void ReceiveFromHelmStation(FromHelmStation payload)
    {
        EnginePower.SetPowerLevel(payload.enginePower, payload.enginePendingPower);
        JumpPower.SetPowerLevel(payload.jumpPower, payload.jumpPendingPower);
        ThrusterPower.SetPowerLevel(payload.thrusterPower, payload.thrusterPendingPower);
        Crew.SetToNumber(payload.crew);
        Jump.SetLampColor(payload.jumpIsAvailable);
        Evac.SetLampColor(payload.crewIsEvac);
        Action1.interactable = payload.action1IsAvailable;
        Action2.interactable = payload.action2IsAvailable;
        Action3.interactable = payload.action3IsAvailable;
        Shields.SetShields(
            payload.maxShields,
            payload.foreShields,
            payload.aftShields,
            payload.portShields,
            payload.starShields
        );
        ReservePower.SetPowerLevel(payload.reserve, payload.maxReserve);
    }

    private void ClickButton(string id)
    {
        var msg = new Message<ButtonClick>()
        {
            c = "button",
            p = new ButtonClick() { id = id },
            e = 0
        };
        Network.Send(msg);
    }

    public void DoAction1()
    {
        ClickButton("helm.action1");
    }

    public void DoAction2()
    {
        ClickButton("helm.action2");
    }

    public void DoAction3()
    {
        ClickButton("helm.action3");
    }

    public void Activate()
    {
        // this is fired when the interface is changed to this component
    }

}

File: GameServer/src/Helm.ts
import { Station } from './Station';

export class FromHelmInterface {
    public throttle: number = 0.0;
    public yaw: number = 0.0;
    public pitch: number = 0.0;
    public enginePower: number = 0;
    public jumpPower: number = 0;
    public thrusterPower: number = 0;
}

export class ToHelmInterface {
    public enginePower: number = 0;
    public enginePendingPower: number = 0;
    public jumpPower: number = 0;
    public jumpPendingPower: number = 0;
    public thrusterPower: number = 0;
    public thrusterPendingPower: number = 0;
    public crew: number = 0;
    public crewIsEvac: boolean = false;
    public jumpIsAvailable: boolean = false;
    public action1IsAvailable: boolean = false;
    public action2IsAvailable: boolean = false;
    public action3IsAvailable: boolean = false;
    public maxShields: number = 0;
    public foreShields: number = 0;
    public aftShields: number = 0;
    public portShields: number = 0;
    public starShields: number = 0;
    public reserve: number = 0;
    public maxReserve: number = 0;
}

export class ToMainViewScreen {
    public speed: number = 0.0;
    public yaw: number = 0.0;
    public pitch: number = 0.0;
}

export class Helm extends Station {
    public get prefix(): string {
        return 'helm';
    }

    public toInterface() {
        var clients = global.server.clients.filter(
            (c) => c.id != 'mainViewScreen'
        );
        var toHelmInterface: ToHelmInterface = {
            enginePower: global.ship.engine.power,
            enginePendingPower: global.ship.engine.pendingPower,
            jumpPower: global.ship.jumpEngine.power,
            jumpPendingPower: global.ship.jumpEngine.pendingPower,
            thrusterPower: global.ship.thrusters.power,
            thrusterPendingPower: global.ship.thrusters.pendingPower,
            crew: 100,
            crewIsEvac: true,
            jumpIsAvailable: false,
            action1IsAvailable: this.action1 ? this.action1.isAvailable : false,
            action2IsAvailable: this.action2 ? this.action2.isAvailable : false,
            action3IsAvailable: this.action3 ? this.action3.isAvailable : false,
            maxShields: global.ship.shields.maxShields,
            foreShields: global.ship.shields.fore,
            aftShields: global.ship.shields.aft,
            portShields: global.ship.shields.port,
            starShields: global.ship.shields.starboard,
            reserve: global.ship.reactor.reserve,
            maxReserve: global.ship.reactor.maxReserve,
        };
        for (var client of clients) {
            global.server.tell(client, 'helm', toHelmInterface);
        }
    }

    public fromInterface(from: FromHelmInterface) {
        // set desired engine power
        if (from.enginePower < 0) {
            // ignore
        } else if (from.enginePower == global.ship.engine.power) {
            global.ship.engine.pendingPower = 0;
        } else {
            global.ship.engine.pendingPower =
                from.enginePower - global.ship.engine.power;
        }

        // set desired jump power
        if (from.jumpPower < 0) {
            // ignore
        } else if (from.jumpPower == global.ship.jumpEngine.power) {
            global.ship.jumpEngine.pendingPower = 0;
        } else {
            global.ship.jumpEngine.pendingPower =
                from.jumpPower - global.ship.jumpEngine.power;
        }

        // set desired thruster power
        if (from.thrusterPower < 0) {
            // ignore
        } else if (from.thrusterPower == global.ship.thrusters.power) {
            global.ship.thrusters.pendingPower = 0;
        } else {
            global.ship.thrusters.pendingPower =
                from.thrusterPower - global.ship.thrusters.power;
        }

        // send the packet to the helm interfaces
        this.toInterface();

        // send the packet to the main view screen
        var mainViewScreen = global.server.clients.find(
            (c) => c.id == 'mainViewScreen'
        );
        if (mainViewScreen) {
            var toMainViewScreen: ToMainViewScreen = {
                speed: from.throttle * global.ship.thrust + global.ship.speed,
                yaw: from.yaw * global.ship.agility * 0.75,
                pitch: from.pitch * global.ship.agility,
            };
            global.logger.info(
                `to: ${mainViewScreen.id} - ${toMainViewScreen.yaw} x ${toMainViewScreen.pitch} @ ${toMainViewScreen.speed}`
            );
            global.server.tell(mainViewScreen, 'helm', toMainViewScreen);
        }
    }
}
